---
title: "SNP densities and sliding windows"
author: "Bradford Condon"
date: "October 20, 2016"
output:
  html_document:
    toc: yes
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Overview

Goal:

Categorize recombination within the genome.

We want to identify SNP-dense regions that are (high SNP within-clade) and (low SNP out-clade).

To do this, I will either use the shell of my sliding window script in perl, or write a new function in R.

Once windows are identified, I want to plot them on the scaffolds, **colored by strain**.

Note: FLTB refers to Festuca,Lolium, Triticum, Bromus.  Any strains grouping in *either* of the wheat blast clades are placed in the FLTB clade for this analysis.  This is desireable for looking for recombination events/ regions that are most similar to non-wheat blast isolates.

##Source data

the loadData.R script reads in SNP report sheets generated by the SNP-finder algorithm developed by Dr. Mark Farman.  This currently points at directories on my local machine and will need to be modified for use elsewhere (and for new SNP reports).

CreateSlidingWindows.R generates the sliding window dataframes.

helperFunctions.R includes all additional functions.

```{r}
#load dependencies etc
library(plyr)
library(ggplot2)
library(knitr)
library(dplyr)
source("R/helperFunctions.R")
```


Load image of results
```{r, eval=TRUE}
#load image created by LoadData.R
#load("IA1andBR80ReadSNPs.RDATA")
#this should be the latest with all - double check
load("IAandBR80andWBKY_SNPsreadv3.RDATA")
```


##Summarizing Br80 scaffolds
First, let's summarize scaffolds based on SNPs by clade.  This will allow me to develop the sorts of methods I'll use once windows are defined

```{r}
#first, sum number of SNPs for each file for each scaffold
snpsGrouped <- summarise(group_by(br80AllSnpDF, scaffoldRef, fileName, Clade), SNPs = n())
#now, take the average for each scaffold
snpsAvg <- summarise(group_by(snpsGrouped, fileName), average = mean(SNPs), SD = sd(SNPs))
kable(snpsAvg)
#Plot this in whiskerplot form

ggplot(snpsGrouped, aes(x= fileName, y = SNPs, fill= Clade)) + geom_boxplot()+ ggtitle("br80 SNPs per scaffold")
#digitaria skews the plot too much: remove

snpsGrouped2 <- filter(snpsGrouped, Clade != 'digitaria')
ggplot(snpsGrouped2, aes(x= fileName, y = SNPs, fill= Clade)) + geom_boxplot()+ scale_y_log10()+ggtitle("Number of SNPs/scaffold against Br80")+ ylab("number of SNPs (log)") +  theme(axis.text.x = element_text(angle = 60, hjust = 1)) + xlab("")


#how many SNPs are there against BR80 and IA1 in each genome?

br80byTaxon<- summarise(group_by(br80AllSnpDF, fileName), SNPs = n())
kable(br80byTaxon)


#kable(group_by(IA1, fileName), SNPs = n())

```

We can see that the FLTB isolates have lower SNPs/scaffold.  Overall this plot isn't that informative because it isn't scaled by scaffold size.
We can transform the number of SNPs into SNPs/kb.

```{r}
#change to character, merge
br80lengths$scaffoldRef<- as.character(br80lengths$scaffold)
snpsGrouped2$scaffoldRef<- as.character(snpsGrouped2$scaffoldRef)
total <- merge(br80lengths,snpsGrouped2,by="scaffoldRef")

totalPercent<- mutate(total, scaledSNPs = SNPs * 100/ ( as.numeric(length))) 
ggplot(totalPercent, aes(x= fileName, y = scaledSNPs, fill= Clade)) + geom_boxplot()+ggtitle("Percent divergence of scaffolds against Br80")+ ylab("Percent divergence (SNPs*100/scaffoldlength") +  theme(axis.text.x = element_text(angle = 60, hjust = 1)) + xlab("")+scale_y_continuous(limits = c(0, 10))


totalPercent<- group_by(totalPercent, Clade)
statsByClade<- summarise(totalPercent, meanSNPs = mean(scaledSNPs), SD = sd(scaledSNPs), medianSNPs = median(scaledSNPs))

kable(statsByClade)

```

What's interesting is that the general scaffold-by-scaffold % SNPs isn't *hugely* variable.  The mean is universally lower for FLTB, but 0.5% as opposed to 0.6-1.0%.  However, the shape of the data shows that the median for FLTB is 0.04 as opposed to 0.27-0.59.   


===

*goal*: examine SNP density in scaffold00050

##10-29-16 different br80 subset
```{r}
dataSubset2Br80 <- (br80AllSnpDF[which(br80AllSnpDF$fileName %in% c("B51")),])
snpsToPlot<- dataSubset2Br80[dataSubset2Br80$scaffoldRef == 'scaffold00050',]
title <- paste("br80 vs B51")
 ggplot(snpsToPlot, aes(x=locRef, fill= fileName)) + geom_histogram(binwidth = 2000)+ ggtitle(title)+xlab("Br80 position")+coord_cartesian(ylim = c(1, 50))

snpsToPlot<- br80AllSnpDF[br80AllSnpDF$scaffoldRef == 'scaffold00050',]
title <- paste("br80 vs all")
ggplot(snpsToPlot, aes(x=locRef, fill= Clade)) + geom_histogram(binwidth = 2000)+ ggtitle("Br80 scaffold 50 v all")+xlab("Br80 position")+coord_cartesian(ylim = c(1, 50))+ facet_wrap(~fileName, drop = TRUE, ncol = 6)
```

11-1-16

Plot specific sacaffolds requested by mark

```{r}
kable(summarise(group_by(br80AllSnpDF, fileName), count=n()))

dataSubsetBr80<- (br80AllSnpDF[which(br80AllSnpDF$fileName %in% c("WBKY", "IC17", "B2", "Br7", "T25", "IA1", "CHRF", "ML33", "SSFL14-3", "B51")),])
dataSubsetBr80$fileName  <- factor(dataSubsetBr80$fileName, levels = c("SSFL14-3", "B2", "B51", "Br7", "T25", "CHRF", "WBKY", "IA1", "IC17", "ML33"))

s37DF <- filter(dataSubsetBr80, scaffoldRef == 'scaffold00037')
s50DF <- filter(dataSubsetBr80, scaffoldRef == 'scaffold00050')


kable(summarise(group_by(s37DF, fileName, Clade), count=n()))



kable(summarise(group_by(filter(br80AllSnpDF, scaffoldRef == 'scaffold00037'), fileName, Clade), count = n()))

kable(summarise(group_by(filter(br80AllSnpDF, scaffoldRef == 'scaffold00050'), fileName, Clade), count = n()))



binSize <- 2000

  ggplot(s37DF, aes(x=locRef, fill= Clade)) + geom_histogram(binwidth = binSize)+ ggtitle("Br80 Scaffold 37")+xlab("Scaffold position")+    facet_wrap(~fileName, drop = TRUE, ncol = 2)+coord_cartesian(ylim = c(0,50))
  ggplot(s50DF, aes(x=locRef, fill= Clade)) + geom_histogram(binwidth = binSize)+ ggtitle("Br80 Scaffold 50")+xlab("Scaffold position")+    facet_wrap(~fileName, drop = TRUE, ncol = 2)+coord_cartesian(ylim = c(0,50))
  
  
  filter(s50DF, fileName == 'Br7')

  
   
```

##Corresponding Br7 scaffolds

We want to look at Scaffold 31 in Br7.

```{r}

dataSubsetBr7<- (br7AllSnpDF[which(br7AllSnpDF$fileName %in% c("WBKY", "IC17", "B2", "Br80", "T25", "IA1", "CHRF", "ML33", "SSFL14-3", "B51")),])

Br7Subset <- filter(dataSubsetBr7, scaffoldRef == 'scaffold00031')
  ggplot(Br7Subset, aes(x=locRef, fill= Clade)) + geom_histogram(binwidth = binSize)+ ggtitle("Br7 Scaffold 31")+xlab("Scaffold position")+    facet_wrap(~fileName, drop = TRUE, ncol = 2)+coord_cartesian(ylim = c(0,50))

  Br737ALL <- filter(br7AllSnpDF, scaffoldRef == 'scaffold00031')

  kable(summarise(group_by(Br737ALL, fileName, Clade), count=n()))

  ggplot(data = (summarise(group_by(Br737ALL, fileName, Clade), count=n())), aes(x=fileName, y=count, fill=Clade)) + geom_bar(stat = "identity")


  
```




PL3-1, FT05-1 and wBKY are SNP dense.  What do those corresponding scaffolds look like in WBKY vs rice?

```{r}
WBKYregion <- filter(br80AllSnpDF, fileName == 'WBKY', scaffoldRef == 'scaffold00050')
#range
min(WBKYregion$locQ)
max(WBKYregion$locQ)

hist(WBKYregion$locQ)
#plot this scaffold v all

snpsToPlot <- filter(WBKYAllSnpDF,  scaffoldRef == 'scaffold00005')
 ggplot(snpsToPlot, aes(x=locRef, fill= Clade)) + geom_histogram(binwidth = 2000)+ ggtitle("WBKY scaffold 5 (corresponding to BR80 scaffold 50) v all")+xlab("WBKY position")+coord_cartesian(ylim = c(1, 50))+ facet_wrap(~fileName, drop = TRUE, ncol = 6)
ggplot(snpsToPlot, aes(x=locRef, fill= Clade)) + geom_histogram(binwidth = 2000)+ ggtitle("WBKY scaffold 5 (corresponding to BR80 scaffold 50), 3.5MB-4MB, v all")+xlab("WBKY position")+coord_cartesian(ylim = c(1, 50), xlim = c(3500000,4000000))+ facet_wrap(~fileName, drop = TRUE, ncol = 6)

```


====

#Sliding windows
Next- we want to create windows, and score those windows based on number of SNPs by clade.  In particular, we want to know

* Max SNPs/window/clade
* Min SNPs/window/clade
* Clade which had min SNP value?

This is done in the createSlidingWindows.R script.

#Analyzing sliding windows
To help with scoring the sliding windows, we need to ask some basic questions about them.

* What is the distribution of number of SNPs per window for each clade?

```{r}
#load sliding windows
load("br80Windows_v3_1kb.robj")

snpsGrouped <- summarise(group_by(br80WindowsDF, Scaffold, fileName, Clade), avgSNPs = mean(count))

ggplot(data = snpsGrouped, aes(x=fileName, y = avgSNPs, fill = Clade))+ geom_boxplot()+ theme(axis.text.x = element_text(angle = 60, hjust = 1)) +ggtitle("Br80, average number of SNPs 1kb/window")

ggplot(data = snpsGrouped, aes(x=avgSNPs))+ geom_histogram()+ggtitle("Br80 reference, log(average number of SNPs) 1kb/window")+facet_wrap(~Clade, nrow = 3)+scale_y_log10()

#just wheat blast

FLTBgrouped <- group_by(filter(br80WindowsDF, Clade == 'FLTB' ), Scaffold, fileName, Window)

ggplot(data = FLTBgrouped, aes(x=fileName, y = count))+geom_boxplot()+ theme(axis.text.x = element_text(angle = 60, hjust = 1)) +ggtitle("Br80, average number of SNPs 1kb/window, FLTB only")

ggplot(data = FLTBgrouped, aes(x=count))+ geom_histogram()+ggtitle("Br80 reference, SNPs/1kb window")+facet_wrap(~fileName, nrow = 3)+scale_x_continuous(limits= c(0, 75))
#note- the above plot doesn't account for windows with n SNPs = 0.

```

This demonstrates that setting min SNPs = 25 (2.5% divergence) for a given *taxon* will seperate the vast majority of windows

#scoring sliding windows

Critical!  Right now the script does the following *for each window region*:

* 1) Windows with 0 FLTB snps are set as being FLTB origin.
* 2) Windows without FLTB/non-FLTB SNPS are ignored.
* 3) Windows are classified based on the FLTB/non-FLTB score.
  * the *AVERAGE* FLTB is compared with the *MINIMUM* non-FLTB.
  * Whichever of the two is lower is considered the source for that window.


Possible big change to make: There is no minimum number of SNPs to consider a scaffold.  If we're interested in SNP-dense regions, want to set some sort of minimum.  But compared to which strains/clades?  All of them?  Just FLTB?

(Need minimum number of SNPs in wheat BLAST)

Issue: sometimes, clades are tied for minimum.  IE brachiara and Oryza will both have only 1 SNP.  what then?

```{R, eval=FALSE}
allScaffoldsTracker<- br80WindowsDF
#let's classify each window as being "outsourced or not".
#I'm going to do define this as a minimum "out" distance being LESS than the average "in" distance
#Final out is going to be
#scaffold - window pos - in/out - number of snps


```


#Plotting windows

```{r, eval=FALSE}

#Convert window character to numeric
dfTracker$window<- as.numeric(gsub( "-.*", "", x= as.character(dfTracker$window)))
##Loop through every scaffold, print to a PDF.
scaffoldList <- as.character((unique(dfTracker$Scaffold)))
pdf("br80_5kwindows.pdf", 8.5, 11)
for (i in scaffoldList){
scaffToPlot<- subset(dfTracker, Scaffold == i)
title <- paste("br80", i)
p<- ggplot(scaffToPlot, aes(x=window, fill = match)) +  ggtitle(title)+xlab("Br80 position")+ geom_histogram() 
plot(p)
}
dev.off()
```

#Plot SNP density to multipage PDF
Working with a different small subset, but including rice, and coloring by clade.

```{r, eval=FALSE}
dataSubsetBr80<- (br80AllSnpDF[which(br80AllSnpDF$fileName %in% c("WBKY", "IC17", "B2", "Br7", "T25", "IA1", "CHRF", "ML33")),])
#refactor for best order
dataSubsetBr80$fileName  <- factor(dataSubsetBr80$fileName, levels = c("B2", "Br7", "T25", "CHRF", "WBKY", "IA1", "IC17", "ML33"))
#plot all SNPs to PDF
PlotSNPstoMultiPDF( dataSubsetBr80, pdfName = "Br80_SNPs_subset.pdf")


##11-2-16
#plot additional br80 with Bm88324 instead of ML33
dataSubsetBr80<- (br80AllSnpDF[which(br80AllSnpDF$fileName %in% c("WBKY", "IC17", "B2", "Br7", "T25", "IA1", "CHRF", "Bm88324")),])
#refactor for best order
dataSubsetBr80$fileName  <- factor(dataSubsetBr80$fileName, levels = c("B2", "Br7", "T25", "CHRF", "WBKY", "IA1", "IC17", "Bm88324"))
#plot all SNPs to PDF
PlotSNPstoMultiPDF( dataSubsetBr80, pdfName = "Br80_SNPs_subset_Bm88324.pdf")

```

Again, with the rice set
```{r, eval=FALSE}
dataSubsetIA1<- (IA1AllSnpDF[which(IA1AllSnpDF$fileName %in% c("WBKY", "IC17",  "Br7", "T25", "IE1K", "Guy11", "CHRF", "ML33")),])
PlotSNPstoMultiPDF(dataSubsetIA1, pdfName = "IA1_SNPs_subset.pdf")
```

#Plotting specific scaffolds:


```{r}

#SSFL02

dataSubsetBr80$fileName  <- factor(dataSubsetBr80$fileName, levels = c("B2", "Br7", "T25", "CHRF", "WBKY", "IA1", "IC17", "ML33", "SSFL02"))
#plot all SNPs to PDF
PlotSNPstoMultiPDF( dataSubsetBr80, pdfName = "Br80_SNPs_subset.pdf")

```


#calculating SNP dispersal

What is the distribution of SNPs?/What percentage of all SNPs are on a given genome?
For SNPs occuring on a given genome, are they usually very close together or spread out?


```{r, summaryStats,eval=TRUE}
dataSubsetBr80<- (br80AllSnpDF[which(br80AllSnpDF$fileName %in% c("WBKY", "IC17", "B2", "Br7", "T25", "IA1", "CHRF", "ML33")),])
#refactor for best order
dataSubsetBr80$fileName  <- factor(dataSubsetBr80$fileName, levels = c("B2", "Br7", "T25", "CHRF", "WBKY", "IA1", "IC17", "ML33"))

groupedBr80df<- group_by(dataSubsetBr80, scaffoldRef, fileName, Clade)
br1kList <- br80lengths[as.numeric(br80lengths$length) > 1000,]
grouped1kBr80<- filter(groupedBr80df, scaffoldRef %in% br1kList$scaffold)
#subset data to scaffolds on the 1k list
#Convert locations to percent
groupedBr801k2<- mutate(grouped1kBr80, percentLoc =(100*locRef/(br80lengths[as.character(br80lengths$scaffold)== scaffoldRef,2])) ) 


sumsBr80<-  summarise(group_by(groupedBr801k2, scaffoldRef, fileName,  Clade),  count = n(), scaffoldSD = round(sd(percentLoc), digits =0))


ggplot(data = sumsBr80, aes(x =  fileName , y = scaffoldSD, fill = Clade ))+ geom_boxplot()+ ggtitle("SD of % SNP location, all scaffolds > 1kb, BR80")

###plot SD vs number of SNPs total


ggplot(data = sumsBr80, aes(x =  log(count) , y = scaffoldSD, color = Clade ))+ geom_point()+ ggtitle("SD of % SNP location vs number of SNPs, all scaffolds > 1kb, BR80")+ facet_grid(~Clade)

#filter sumsBr80 to those with at least 1,000 SNPs

ggplot(data = filter(sumsBr80, count >= 500), aes(x =  fileName , y = scaffoldSD, fill = Clade ))+ geom_boxplot()+ ggtitle("SD of % SNP location, all scaffolds > 1kb and with > 500 SNPs, BR80")

```

repeat with IA
```{r}
#sum SNPs for each filename for each scaffoldRef
groupedIAdf<- group_by(IA1AllSnpDF, scaffoldRef, fileName)
IA1kList <- IA1lengths[as.numeric(IA1lengths$Length) > 1000,]
grouped1kIA1<- filter(groupedIAdf, scaffoldRef %in% IA1kList$ID)
#subset data to scaffolds on the 1k list
#Convert locations to percent
grouped1kIA1<- mutate(grouped1kIA1, percentLoc =(100*locRef/(IA1lengths[as.character(IA1lengths$ID)== scaffoldRef,2])) ) 


sumsIA1<-  summarise(group_by(grouped1kIA1, scaffoldRef, fileName,  Clade),  count = n(), scaffoldSD = round(sd(percentLoc), digits =0))
levels(factor(sumsIA1$fileName))

#relevel
sumsIA1$fileName  <- factor(sumsIA1$fileName, levels = c("Br7", "T25", "CHRF", "WBKY", "Guy11", "IC17", "IE1K", "ML33"))

ggplot(data = sumsIA1, aes(x =  fileName , y = scaffoldSD, fill = Clade ))+ geom_boxplot()+ ggtitle("SD of % SNP location, all scaffolds > 1kb, IA1")


```


#Calculating SNP disperal V2:  using nearest neighbor 10-26-16 

Approach taken- calculate distance to previous/next SNP, pick the minimum of these.

* Sort DF by scaffold name and location (essential for lag and lead to be checking the correct SNPs)
* lag() and lead()  calculate based on previous/next
* pmin necessary instead of min (because column values?).  


```{r, nearest SNP neighbor}
#sort by scaffoldRef and then locRef, then group by scaffoldref/fileName/clade, then calculate nearest neighbor

IA1neighbor<- mutate(group_by( arrange(IA1AllSnpDF,fileName,scaffoldRef, locRef), scaffoldRef, fileName, Clade), prevNeighbor = as.numeric(locRef - lag(locRef)), nextNeighbor = as.numeric(lead(locRef)- locRef), nearNeighbor = pmin(prevNeighbor, nextNeighbor, na.rm=TRUE) )
head(IA1neighbor)
IA1neighbor<- filter(IA1neighbor, Clade != "")



ggplot(data = IA1neighbor, aes(x =  nearNeighbor), color = fileName)+ geom_histogram()+ ggtitle("SNP nearest neighbor, IA1 reference")+ ylab("density")+facet_wrap(~Clade, nrow= 3 )+scale_x_log10()+xlab("nearest neighbor of a given SNP (log scale)")

#remove blank classified
IA1neighbor<- filter(IA1neighbor, Clade != "")

IA1neighborSum<-  summarise(group_by(IA1neighbor,  Clade),  count = n(), NNmean = round(mean(nearNeighbor, na.rm = TRUE), digits =0), NNSD = round(sd(nearNeighbor, na.rm = TRUE), digits =0))

kable(IA1neighborSum)

IA1scafAvg <- summarise(group_by(IA1neighbor, Clade, scaffoldRef), meanNeighborDist = mean(nearNeighbor))

ggplot(data = IA1scafAvg, aes(x =  Clade , y = meanNeighborDist, fill = Clade ))+ geom_boxplot()+ ggtitle("Average SNP nearest neighbor, IA1 reference")+ ylab("Average Closest SNP distance")+ ylim(0, 5000)

#repeat for Br80
#

br80neighbor<- mutate(group_by( arrange(br80AllSnpDF,fileName,scaffoldRef, locRef), scaffoldRef, fileName, Clade), prevNeighbor = as.numeric(locRef - lag(locRef)), nextNeighbor = as.numeric(lead(locRef)- locRef), nearNeighbor = pmin(prevNeighbor, nextNeighbor, na.rm=TRUE) )

#remove blank classified
br80neighbor<- filter(br80neighbor, Clade != "")

#dont want to avg by scaffold...
ggplot(data = br80neighbor, aes(x =  nearNeighbor))+ geom_histogram()+ ggtitle("SNP nearest neighbor, Br80 reference")+ ylab("density")+facet_wrap(~Clade, nrow= 3 )+scale_x_log10()+xlab("nearest neighbor of a given SNP (log scale)")


br80neighborSum<-  summarise(group_by(br80neighbor,  Clade),  count = n(), NNmean = round(mean(nearNeighbor, na.rm = TRUE), digits =0), NNSD = round(sd(nearNeighbor, na.rm = TRUE), digits =0))

kable(br80neighborSum)

br80scafAvg <- summarise(group_by(br80neighbor, Clade, scaffoldRef), meanNeighborDist = mean(nearNeighbor))


ggplot(data = br80scafAvg, aes(x =  Clade , y = meanNeighborDist, fill = Clade ))+ geom_boxplot()+ ggtitle("SNP nearest neighbor (average/scaffold), Br80 reference")+ ylab("Average closest SNP distance")+ylim(0,5000)

###what if we only take the scaffolds with >500 SNPs?  


```



##Experimenting with plot types

###Plotting all

```{r, allSamplesTogether}
head(IA1AllSnpDF)
i <- "IA1_scaffold00069"
snpsToPlot<- IA1AllSnpDF[IA1AllSnpDF$scaffoldRef == i,]
title <- paste("IA1", i, "experimental plot, freqpoly")
 ggplot(snpsToPlot, aes(x=locRef, color= fileName)) + geom_freqpoly(binwidth = 20)+ ggtitle(title)+xlab("IA1 position")+ facet_wrap(~Clade, drop = TRUE, ncol = 2)+ coord_cartesian(ylim = c(1, 10))+ guides(color=FALSE)
 
```
